subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/timeout/IntegrationTestTimeoutInterceptor.groovy

```diff
@@ -0,0 +1,180 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.fixtures.timeout
+
+import groovy.transform.EqualsAndHashCode
+import groovy.transform.ToString
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.executer.InProcessGradleExecuter
+import org.gradle.integtests.fixtures.executer.IntegrationTestBuildContext
+import org.gradle.internal.os.OperatingSystem
+import org.spockframework.runtime.SpockAssertionError
+import org.spockframework.runtime.SpockTimeoutError
+import org.spockframework.runtime.extension.IMethodInvocation
+import org.spockframework.runtime.extension.builtin.TimeoutInterceptor
+
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+
+class IntegrationTestTimeoutInterceptor extends TimeoutInterceptor {
+    private static final Pattern UNIX_JAVA_COMMAND_PATTERN = ~/(?i)([^\s]+\/bin\/java)/
+    private static final Pattern WINDOWS_JAVA_COMMAND_PATTERN = ~/(?i)^"(.*[\/\\]bin[\/\\]java\.exe)"/
+    private static final Pattern WINDOWS_PID_PATTERN = ~/([0-9]+)\s*$/
+    private static final Pattern UNIX_PID_PATTERN = ~/([0-9]+)/
+
+    IntegrationTestTimeoutInterceptor(IntegrationTestTimeout timeout) {
+        super(new TimeoutAdapter(timeout))
+    }
+
+    @Override
+    void intercept(final IMethodInvocation invocation) throws Throwable {
+        try {
+            super.intercept(invocation)
+        } catch (SpockTimeoutError e) {
+            Object instance = invocation.getInstance()
+            if (instance instanceof AbstractIntegrationSpec) {
+                String allThreadStackTraces = getAllThreads((AbstractIntegrationSpec) instance)
+                throw new SpockAssertionError(allThreadStackTraces, e)
+            } else {
+                throw e
+            }
+        } catch (Throwable t) {
+            throw t
+        }
+    }
+
+    static String getAllThreads(AbstractIntegrationSpec spec) {
+        try {
+            if (spec.executer.gradleExecuter.class == InProcessGradleExecuter) {
+                return getAllThreadsInCurrentJVM()
+            } else {
+                return getAllThreadsByJstack()
+            }
+        } catch (Throwable e) {
+            def stream = new ByteArrayOutputStream()
+            e.printStackTrace(new PrintStream(stream))
+            return stream.toString()
+        }
+    }
+
+    @EqualsAndHashCode
+    @ToString
+    static class JavaProcessInfo {
+        String pid
+        String javaCommand
+
+        String getJstackCommand() {
```

Comment:
You could use `OperatingSystem.getExecutableName(...)` or maybe even better `Jvm.getExecutable`. WDYT?

Reply:
Thanks for the tips, I wasn't aware of these two methods. However, aren't these two methods only getting `executable` of current JVM? What if I'm trying to run `jstack` on another JVM with different Java version?